### PR TITLE
fix(ICC/Deathbringer Saurfang) NPC Muradin Dialogues

### PR DESCRIPTION
**Pull request by pangolp**
## Changes Proposed:

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/8921
- Closes https://github.com/azerothcore/azerothcore-wotlk/pull/13506

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.

WoWScrnShot_052024_041835

WoWScrnShot_052024_041928

## How to Test the Changes:

1. `.tele IcecrownCitadel`
2. Eliminate the first 3 bosses, and then start the fourth fight. Without the pull request, the script is not updated, and the dialogs are displayed incorrectly. I don't know if there is another way to do it, but I looked for thousands of ways, and the only one that occurred to me, as a final point, to fix it, until a better way appears, was this.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by PkllonG** [comment]
https://github.com/azerothcore/azerothcore-wotlk/pull/13506

**Comment by pangolp** [comment]
> https://github.com/azerothcore/azerothcore-wotlk/pull/13506
> You try this and it works fine on my server.

I'll try it. I don't understand why they closed their pull request. I've been trying to figure this out for quite some time. I made many changes, before uploading this PR. Since I couldn't find a solution, I leaned towards this path.

**Comment by pangolp** [comment]
@PkllonG Could you please upload an image of the `.npc info` command from your server? Thank you. I have uploaded images of the results of mine. However, I would like to see the result you get.

**Comment by PkllonG** [comment]
You try, it's definitely good.

**Comment by PkllonG** [comment]
image

**Comment by pangolp** [comment]
It seems like your change would be correct, I'll try it later. Thank you.

**Comment by pangolp** [comment]
Your code seems to be correct. I only made a small modification. The `GetScriptId` was added differently within the `UpdateEntry`.

**Comment by pangolp** [comment]
I just made the change I mentioned, if you want, you can try it to see if you get the same result. I just tried it. It seems to work. We can also close this pull request, and re-open yours. I gave him credits for the commit. But I leave it to your discretion. It's a shame not to have seen it before. I have been trying to solve this for several months and carried out tests of all kinds.

**Comment by PkllonG** [comment]
You can come,

File: src/server/scripts/Northrend/IcecrownCitadel/instance_icecrown_citadel.cpp
Comment (PkllonG) [inline src/server/scripts/Northrend/IcecrownCitadel/instance_icecrown_citadel.cpp:406]: Don't delete here

File: src/server/game/Entities/Creature/Creature.cpp
Comment (PkllonG) [inline src/server/game/Entities/Creature/Creature.cpp:597]: Please delete here

**Comment by pangolp** [comment]
@PkllonG do you want to make those changes if the result is the same? Did you try it? Does it affect something I'm not considering? I need context because you're asking me for those changes. If it affects something that doesn't, then it would be good to know.

**Comment by PkllonG** [comment]
The result is the same

**Comment by pangolp** [comment]
> The result is the same

All the more reason, if the result is the same, when you consult the script inside the UpdateEntry, you save yourself from doing it outside it. Unless it affects another mechanic, I think it would be better that way.

Unless we're making a mistake, and then yes, we can do it the way you indicate. But if the result is the same, I prefer to do it this way.

**Comment by pangolp** [comment]
If I have to make those changes, the truth is, I don't care, I have no problem doing it. What I want is for it to solve the problem. I don't care about the method, I'm not interested in being right. I'm just looking for the fault to be corrected.

**Comment by pangolp** [comment]
Restore how to update the script. Actually, as long as the problem is solved, it is indistinct to me, if it is updated in creature or in the instance. I still believe that it is better within UpdateEntry because I would always update the script of the npc, but to avoid confessions, I left it as it was.

**Comment by pangolp** [comment]
https://github.com/azerothcore/azerothcore-wotlk/assets/2810187/0bbd5e9b-6bc8-4957-b765-ae0e25498edf

https://github.com/azerothcore/azerothcore-wotlk/assets/2810187/d8988465-0271-4ded-a180-751743fa76d5

**Comment by PkllonG** [comment]
Not all UpdateEntry need to be loaded with AI,
Changing only partial code makes it easier to merge code and safer.

**Review by PkllonG — COMMENTED** [review]
It's great now

**Review by PkllonG — COMMENTED** [review]
Good job

**Comment by pangolp** [comment]
> Good job

It was your job. My code was different. But if that way works, and we don't have to add a spawn, it is better to do it this way.

**Comment by PkllonG** [comment]
It has been working on my server since very early days and has not found bugs in a few years

File: src/server/game/Entities/Creature/Creature.cpp
Comment (Nyeriah) [inline src/server/game/Entities/Creature/Creature.cpp:3020]: Please explain this change

Comment (pangolp) [reply src/server/game/Entities/Creature/Creature.cpp:3020]: From what I could see, through this change, what can be achieved is that if the npc has an updateEntry, then it takes the script that corresponds to it, otherwise, it always takes the script from the npc before the update. That's why the dialogues worked poorly. Initially, I had proposed something else, which was to spawn the npc, but this is not the first one to have this problem. What is not entirely clear to me, if necessary, put a true in the `bool changelevel`, which is the third parameter of the method.
```cpp
bool Creature::UpdateEntry(uint32 Entry, const CreatureData* data, bool changelevel, bool updateAI)
```

**Comment by pangolp** [comment]
In the image of the pull request, you can see how the npc updates its entry and also the id, as the name of the script it should use. Currently, without this change, muradin, despite doing the updateEntry, continues to retain the id of the original npc, as well as its script.

**Comment by pangolp** [comment]
Currently, in the master branch of the repository, the npc, when doing an updateEntry, updates several of the data, but does not apply changes to id1 and consequently, does not update the script it should use either. That is why the dialogue is broken, and for that reason, this pull request is being created to solve that problem.

muradin

After applying the changes, you can confirm that the id1 of the npc is updated, and also its script. If this pull request is approved, we may have to review several more npc. Mainly, in those instances, where updateEntry occur.

muradin_fix

File: src/server/scripts/Northrend/IcecrownCitadel/instance_icecrown_citadel.cpp
Comment (Nyeriah) [inline src/server/scripts/Northrend/IcecrownCitadel/instance_icecrown_citadel.cpp:404]:
```suggestion
                        creature->UpdateEntry(NPC_SE_MURADIN_BRONZEBEARD, true);
```

File: src/server/game/Entities/Creature/Creature.cpp
Comment (pangolp) [reply src/server/game/Entities/Creature/Creature.cpp:3020]: How else do we solve the problem? Because if we set that value in True, the script is not updated and there is a flaw that can be seen in the videos. I uploaded several videos explaining the difference when applying the changes, but if there's another way to solve the problem, I'd like to know. The method might not be right, it can be changed. The goal is to correct the problem.

Comment (pangolp) [reply src/server/game/Entities/Creature/Creature.cpp:3020]: From the mobile phone, I didn't see that you had sent a suggestion for change. I applied it, I'll try it. I'll see if I get the same result. Thanks.

**Comment by pangolp** [comment]
@Nyeriah after the change is applied, mainly in the `Creature.cpp` file, I will continue updating several npcs, from other instances, or within this same one, that have the same problem. But first, we have to test if this works before we can make the other changes.

**Comment by pangolp** [comment]
If necessary, we can split the issue into 2 pull requests. One to update the changes in Creature.cpp and another to make the change to the script. In case it needs to be reverted, I have no problem creating a second pull request if necessary and splitting the issue. If you need anything, let me know.

**Comment by pangolp** [comment]
I've been testing the changes, and there seems to be no difference. The script works as it should. However, the `id1` inside the parenthesis is not updated. It is the only value that remains unchanged, and is still the id of the original npc.

`Current Entry: 37200 of (37187, 0, 0).`

https://www.wowhead.com/wotlk/npc=37200/muradin-bronzebeard
https://www.wowhead.com/wotlk/npc=37187/high-overlord-saurfang

I don't know if we should update that as well, or if it's not necessary to do so. In the script, no changes are made on that subject.

**Comment by Nyeriah** [comment]
I'll look into this over the weekend.

**Comment by pangolp** [comment]
> I'll look into this over the weekend.

Okay. Look into it and let me know if you have any questions, so I can modify this pull request and after it is merged, create another one with the changes I mentioned, because I feel that it won't be the only NPC that we have to make this type of change to. There may be more within the instance and in other instances as well. For example, in Pit of Saron

**Comment by PkllonG** [comment]
Maybe I can help a little because the relevant changes have been working fine on my server for a long time. Bugs have all been fixed.

**Comment by pangolp** [comment]
> Maybe I can help a little because the relevant changes have been working fine on my server for a long time. Bugs have all been fixed.

We could only make the changes to the Creature.cpp file in this PR and then create another PR with the changes to the npc that we must update. So, if we were to reverse one, it wouldn't affect the other.

**Review by Nyeriah — APPROVED** [review]

**Comment by pangolp** [comment]
Thanks.

**Pull request merged**
